composer/loggers/mosaicml_logger.py

```diff
@@ -0,0 +1,174 @@
+
+        if os.environ.get(_MOSAICML_API_KEY_ENV, None) is None:
+            self.skip_logging = True
+            warnings.warn(
+                f"No api_key set for environment variable {_MOSAICML_API_KEY_ENV}. MosaicML logger will be a no-op.")
+
+        self.buffered_data = []
+        self.flush_every_n_batches = flush_every_n_batches
+        self.max_logs_in_buffer = max_logs_in_buffer
+
+        self.queue = Queue()
+        self.thread = Thread(target=self._listen_to_queue)
+
+    def _log_metric(self, epoch: int, step: int, log_level: LogLevel, data: TLogData):
+        del log_level  # unused
+
+        if self.skip_logging:
+            return
+
+        log_data = {
+            "job_id": self.job_id,
+            "sweep_id": self.sweep_id,
+            "epoch": epoch,
+            "step": step,
+            "data": data,
+        }
+        self.buffered_data.append(log_data)
+        if len(self.buffered_data) > self.max_logs_in_buffer:
+            self._flush_buffered_data()
+
+    def _training_start(self, state: State, logger: Logger) -> None:
+        del state, logger  # unused
+
+        if self.skip_logging:
+            return
+
+        log.info("Starting MosaicML logger thread.")
+
+        # Start the logging thread
+        self.thread.start()
+
+    def batch_end(self, state: State, logger: Logger):
+        del logger  # unused
+
+        if self.skip_logging:
+            return
+
+        if (state.step + 1) % self.flush_every_n_batches == 0:
+            self._flush_buffered_data()
+
+    def training_end(self, state: State, logger: Logger):
+        del state, logger  # unused
+
+        if self.skip_logging:
+            return
+
+        # Flush any remaining logs on training end
+        self._flush_buffered_data()
+
+        self.queue.put_nowait(_STOP_LOGGING_SIGNAL)
+        self.thread.join()
+
+        log.info(f"MosaicML logger thread has exited.")
+
+    def state_dict(self) -> StateDict:
+        # Storing these fields in the state dict to support run resuming in the future.
+        return {"job_id": self.job_id, "sweep_id": self.sweep_id}
+
+    def load_state_dict(self, state: StateDict) -> None:
+        self.job_id = state["job_id"]
+        self.sweep_id = state["sweep_id"]
+
+    def _flush_buffered_data(self):
+        if len(self.buffered_data) == 0:
+            return
+
+        data_to_write = self.buffered_data.copy()
+        self.buffered_data = []
+
+        self.queue.put_nowait(data_to_write)
+
+    def _listen_to_queue(self):
+        while True:
+            data = self.queue.get(block=True)
+            if data == _STOP_LOGGING_SIGNAL:
+                log.info("MosaicML logger thread received stop logging signal.")
+                self.queue.task_done()
+                return
+
+            try:
+                response = _send_data(job_id=self.job_id, sweep_id=self.sweep_id, data=data)  # type: ignore
+                if response.status_code != 200:
```

Comment:
Ah ok, I would move this into the _send_data function.
But should we ignore errors?

Reply:
Decided to ignore errors for now because custom error handling introduces a lot of code complexity and it's not critical that all logs are uploaded anyway (the only use case so far is benchmarking and the cost of retrying a benchmarking job in the case where the final throughput isn't uploaded is low)